allensdk/brain_observatory/behavior/stimulus_processing.py

```diff
@@ -94,48 +117,153 @@ def get_images_dict(pkl):
     return images_dict
 
 
-def get_stimulus_templates(pkl):
+def get_stimulus_templates(pkl) -> Dict:
+    """
+    Gets dictionary of images presented during experimentation
+    Parameters
+    ----------
+    pkl: pkl file containing the data for the presented stimuli
+
+    Returns
+    -------
+    Dict:
+        Dictionary of images that were presented during the experiment
 
+    """
     images = get_images_dict(pkl)
     image_set_filename = convert_filepath_caseinsensitive(images['metadata']['image_set'])
     return {IMAGE_SETS_REV[image_set_filename]: np.array(images['images'])}
 
 
-def get_stimulus_metadata(pkl):
+def get_stimulus_metadata(pkl) -> pd.DataFrame:
+    """
+    Gets the stimulus metadata for each type of stimulus presented during
+    the experiment. The metadata is return for gratings, images, and omitted
+    stimuli.
+    Parameters
+    ----------
+    pkl: the pkl file containing the information about what stimuli were
+         presented during the experiment
+
+    Returns
+    -------
+    pd.DataFrame:
+        The dataframe containing a row for every stimulus that was presented
+        during the experiment. The row contains the following data,
+        image_category, image_name, image_set, phase, frequency,
+        and image index.
 
-    images = get_images_dict(pkl)
-    stimulus_index_df = pd.DataFrame(images['image_attributes'])
-    image_set_filename = convert_filepath_caseinsensitive(images['metadata']['image_set'])
-    stimulus_index_df['image_set'] = IMAGE_SETS_REV[image_set_filename]
+    """
+    stimuli = pkl['items']['behavior']['stimuli']
+    if 'images' in stimuli:
+        images = get_images_dict(pkl)
+        stimulus_index_df = pd.DataFrame(images['image_attributes'])
+        image_set_filename = convert_filepath_caseinsensitive(images['metadata']['image_set'])
+        stimulus_index_df['image_set'] = IMAGE_SETS_REV[image_set_filename]
+    else:
+        stimulus_index_df = pd.DataFrame(columns=[
+            'image_name', 'image_category', 'image_set', 'phase',
+            'frequency', 'image_index'])
+
+    # if grating are in the pkl add an entry for each grating possible
+    if 'grating' in stimuli:
+        phase = stimuli['grating']['phase']
+        correct_freq = stimuli['grating']['correct_freq']
+        start_idx = len(stimulus_index_df)
+        grating_df = {'image_category': ['grating']*4,
```

Comment:
If this is hard-coded how will we know if additional gratings are added?

It it not possible to get this information from the file?

Reply:
I separated this into another function where it looks for the unique orientation values to remove the hard coded values.